staging/src/k8s.io/kube-aggregator/pkg/apiserver/openapi_aggregator.go

```diff
@@ -173,98 +240,81 @@ func (s *openAPIAggregator) updateOpenAPISpec() error {
 	return s.openAPIService.UpdateSpec(specToServe)
 }
 
-// inMemoryResponseWriter is a http.Writer that keep the response in memory.
-type inMemoryResponseWriter struct {
-	header   http.Header
-	respCode int
-	data     []byte
-}
-
-func newInMemoryResponseWriter() *inMemoryResponseWriter {
-	return &inMemoryResponseWriter{header: http.Header{}}
-}
-
-func (r *inMemoryResponseWriter) Header() http.Header {
-	return r.header
-}
-
-func (r *inMemoryResponseWriter) WriteHeader(code int) {
-	r.respCode = code
+// tryUpdatingServiceSpecs tries updating openAPISpecs map, and keep the map intact if the update fails.
+// note that map's elements are pointers, so the update method should not change them.
+func (s *openAPIAggregator) tryUpdatingServiceSpecs(updateFunc func() error) error {
+	// Keep a backup value and restore it if aggregation failed.
+	backupMap := s.openAPISpecs
```

Comment:
This does not work: you only copy the pointer to the map. You have to iterate through it and copy each key-value pair.

Reply:
what about `backupMap := *s.openAPISpecs`. The thing is updateFunc should not mutate items in the map, it can only creates a new one or remove one.

Reply:
Propose that for Go 2.0. I like it :)

Reply:
fixed in next commit.